lib/mllp/client.ex

```diff
@@ -356,282 +381,384 @@ defmodule MLLP.Client do
+  end
 
-      {:error, _} = err ->
-        err
+  def receiving({:call, _from} = event_kind, request, data)
+      when request in [:is_connected, :reconnect] do
+    connected(event_kind, request, data)
+  end
+
+  def receiving(event, unknown, _data) do
+    unexpected_message(:receiving, event, unknown)
+  end
+
+  ########################################
+  ### End of :gen_statem callbacks ###
+  ########################################
+
+  defp unexpected_message(state, event, message) do
+    Logger.warn(
+      "Event: #{inspect(event)} in state #{state}. Unknown message received => #{inspect(message)}"
+    )
+
+    :keep_state_and_data
+  end
+
+  ## Handle the (fragmented) responses to `send` request from a caller
+
+  defp handle_received(_reply, %{caller: nil} = data) do
+    ## No caller, ignore
+    data
+  end
+
+  defp handle_received(<<@header, _rest::binary>> = reply, data) do
+    receive_impl(reply, data)
+  end
+
+  ## No header in the first packet
+  defp handle_received(_reply, %{receive_buffer: []} = data) do
+    reply_to_caller({:error, :invalid_reply}, data)
+  end
+
+  ## The rest of MLLP (after the header was received)
+  defp handle_received(reply, data) do
+    receive_impl(reply, data)
+  end
+
+  def receive_impl(reply, %{receive_buffer: buffer, last_byte_received: last_byte} = data) do
+    new_buf = update_receive_buffer(buffer, reply)
+
+    case trailer_check(reply, last_byte) do
+      :data_after_trailer ->
+        Logger.error("Client #{inspect(self())} received data following the trailer")
+        reply_to_caller({:error, :data_after_trailer}, data)
+
+      true ->
+        Logger.debug("Client #{inspect(self())} received a full MLLP!")
+        reply_to_caller({:ok, IO.iodata_to_binary(new_buf)}, data)
+
+      false ->
+        Logger.debug("Client #{inspect(self())} received a MLLP fragment: #{reply}")
+
+        data
+        |> Map.put(:receive_buffer, new_buf)
+        |> Map.put(:last_byte_received, String.last(reply))
     end
   end
 
-  defp do_recv(state, length, :infinity) do
-    res = state.tcp.recv(state.socket, length, :infinity)
-    {res, 0}
+  ## "Split trailer" case
+  ##
+
+  defp trailer_check(<<@cr, rest::binary>>, @eb) do
+    if rest == "" do
+      true
+    else
+      :data_after_trailer
+    end
   end
 
-  defp do_recv(state, length, timeout) do
-    timeout_in_ms = System.convert_time_unit(timeout, :microsecond, :millisecond)
-    t1 = System.monotonic_time(:microsecond)
-    res = state.tcp.recv(state.socket, length, timeout_in_ms)
-    t2 = System.monotonic_time(:microsecond)
-    {res, t2 - t1}
+  defp trailer_check(packet, _last_buffer_byte) do
+    case :binary.match(packet, @trailer) do
+      :nomatch -> false
+      {pos, @trailer_length} when pos + @trailer_length == byte_size(packet) -> true
+      _ -> :data_after_trailer
+    end
   end
 
-  defp update_recv_time_credit(:infinity, _), do: {:infinity, 0}
+  defp update_receive_buffer(buffer, packet) do
```

Comment:
This is actually beautiful in that you're creating an improper list, which is an optimization ❤️